File: client-scalaz/src/main/scala/org/openapitools/client/api/OfferLocation.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import OfferLocation._

case class OfferLocation (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
latitude: Option[Double],
longitude: Option[Double],
locationDescription: Option[String],
offer: Option[Offer],
location: Option[RetailerLocation],
distance: Option[Double],
altitude: Option[Double],
customValue: Option[Double],
locationDetail: Option[String],
audiences: Option[List[Audience]],
qrInviteCode: Option[Asset],
notes: Option[List[Note]],
noteCount: Option[Long],
searchIndexUpdated: Option[OffsetDateTime],
inSearchIndex: Option[Boolean],
favoriteCount: Option[Long],
addedCount: Option[Integer],
display: Option[String],
appKey: Option[String],
devicePower: Option[Double],
powerLoss: Option[Double],
udid: Option[String],
billableId: Option[Long],
subType: Option[String],
contentName: Option[String],
contentAsset: Option[Asset],
owner: Option[Account],
secondaryType: Option[String],
categoryIds: Option[List[Long]],
filterIds: Option[List[Long]])

object OfferLocation {
  import DateTimeCodecs._

  implicit val OfferLocationCodecJson: CodecJson[OfferLocation] = CodecJson.derive[OfferLocation]
  implicit val OfferLocationDecoder: EntityDecoder[OfferLocation] = jsonOf[OfferLocation]
  implicit val OfferLocationEncoder: EntityEncoder[OfferLocation] = jsonEncoderOf[OfferLocation]
}

File: client-clojure/src/sirqul_io_t_platform/specs/offer_location.clj
(ns sirqul-io-t-platform.specs.offer-location
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.offer :refer :all]
            [sirqul-io-t-platform.specs.retailer-location :refer :all]
            [sirqul-io-t-platform.specs.audience :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.note :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            )
  (:import (java.io File)))


(def offer-location-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :latitude) float?
   (ds/opt :longitude) float?
   (ds/opt :locationDescription) string?
   (ds/opt :offer) offer-spec
   (ds/opt :location) retailer-location-spec
   (ds/opt :distance) float?
   (ds/opt :altitude) float?
   (ds/opt :customValue) float?
   (ds/opt :locationDetail) string?
   (ds/opt :audiences) (s/coll-of audience-spec)
   (ds/opt :qrInviteCode) asset-spec
   (ds/opt :notes) (s/coll-of note-spec)
   (ds/opt :noteCount) int?
   (ds/opt :searchIndexUpdated) inst?
   (ds/opt :inSearchIndex) boolean?
   (ds/opt :favoriteCount) int?
   (ds/opt :addedCount) int?
   (ds/opt :display) string?
   (ds/opt :appKey) string?
   (ds/opt :devicePower) float?
   (ds/opt :powerLoss) float?
   (ds/opt :udid) string?
   (ds/opt :billableId) int?
   (ds/opt :subType) string?
   (ds/opt :contentName) string?
   (ds/opt :contentAsset) asset-spec
   (ds/opt :owner) account-spec
   (ds/opt :secondaryType) string?
   (ds/opt :categoryIds) (s/coll-of int?)
   (ds/opt :filterIds) (s/coll-of int?)
   })

(def offer-location-spec
  (ds/spec
    {:name ::offer-location
     :spec offer-location-data}))
